Allow non-pesistent topics to be retrieved along with persistent ones from the "GetTopicsOfNamespace" method

gordeevbr opened the pull request:
### Motivation

Please see issue #2009 for a detailed bug report.

In our use case we require using java client with pattern subscription to read from a set of non-persistent topics. Unfortunately, right now this feature doesn't work. After researching the cause for this I have found out that under the hood the client is requesting a list of topics by namespace from the server and then filters them out by pattern and subscribes to them. The method in Pulsar broker NamespaceService class that is responsible for searching for required topics only uses ledgers, thus returning only persistent topics to the client. The goal of this pull request is to provide a solution for that problem.

### Modifications

This pull request updates `getListOfTopics` method of NamespaceService class to also include active non-pesistent topics from local broker cache inside the `multiLayerTopicsMap` collection of BrokerService in the result.

### Result

As a result, requesting a list of topics by namespace using the HTTP API or binary API (and thus via the clients) will add non-persistent topics to search result, allowing pattern subscription to be used with non-persistent topics.

### Considerations

1. Since this method pulls non-persistent topics from local broker cache, this probably means that this solution will only work for Pulsar installations with a single broker. And if there are multiple brokers, results might be inconsistent. Unfortunately I don't really know if non-persistent themselves work in multi-broker setups. I have recently asked on Slack if non-persistent topics are being replicated in any way and @merlimat's response was that they don't. Also it seems to be that some other methods that are working with non-persistent topics are using this very same collection.

2. It seems to me that unit tests have made sure that Java client can work with this setup, but this might still be a breaking change for other clients or if applications working with this API are not expecting non-persistent topics in result.

3. I have made sure that old unit tests inside the `pulsar-broker` subproject are still working and updated some old tests for this particular use case. Are there any more tests that I can add.

Overall, we really need this and I would appreciate if maintainers could share their opinion. Thanks in advance.

sijie commented:
@jiazhai please review this when you have time.

gordeevbr commented:
Just noticed integration tests are failing. Is this my fault? Should I fix that?

srkukarni commented:
retest this please

merlimat commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java:
```diff
@@ -825,6 +827,20 @@ public ServiceUnitId getServiceUnitId(TopicName topicName) throws Exception {
         }
 
+        // Non-persistent topics don't have managed ledges so we have to retrieve them from local cache.
+		synchronized (pulsar.getBrokerService().getMultiLayerTopicMap()) {
```
This will only answer with the non-persistent topics served by the broker that it's handling the particular request. However, topics for a single namespace can be spread across multiple brokers. 

We would need to do something like this ( https://github.com/apache/incubator-pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/NonPersistentTopics.java#L170 ) to get the correct list of non-persistent topics.

I think the best option is to actually use the same API to get the list of non-persistent topics through HTTP. `GET /admin/v2/non-persistent/{tenant}/{namespace}`

merlimat commented:
> Unfortunately I don't really know if non-persistent themselves work in multi-broker setups. I have recently asked on Slack if non-persistent topics are being replicated in any way and @merlimat's response was that they don't. Also it seems to be that some other methods that are working with non-persistent topics are using this very same collection.

 * A single topic (or partition of a topic) is assigned to a single broker, at a given point in time. 
 * Topics belonging to a namespace can be assigned to different brokers (that is true for both persistent and non-persistent topics)
 * Persistent topics use BookKeeper to persist data and maintain state --> When a broker crashes, the topic can move immediately to a different broker and restart with the same exact state.
 * Non-persistent topics do not store any kind of state durably. The messages are kept for a very short amount of time in memory in the broker and dropped if they cannot be delivered immediately. The data for a single non-persistent topic is not stored in multiple machines. If the broker crashes, some messages will be lost and the topic will continue on the new broker.

jiazhai commented on pulsar-broker/src/test/java/org/apache/pulsar/client/impl/PatternTopicsConsumerImplTest.java:
```diff
@@ -74,7 +74,8 @@ public void testPatternTopicsSubscribeWithBuilderFail() throws Exception {
         final String topicName2 = "persistent://my-property/my-ns/topic-2-" + key;
         final String topicName3 = "persistent://my-property/my-ns/topic-3-" + key;
-        List<String> topicNames = Lists.newArrayList(topicName1, topicName2, topicName3);
+		final String topicName4 = "non-persistent://my-property/my-ns/topic-4-" + key;
```
seems some space alignment issue.

jiazhai commented on pulsar-broker/src/test/java/org/apache/pulsar/client/impl/PatternTopicsConsumerImplTest.java:
```diff
@@ -133,6 +134,7 @@ public void testBinaryProtoToGetTopicsOfNamespace() throws Exception {
         String topicName1 = "persistent://my-property/my-ns/pattern-topic-1-" + key;
         String topicName2 = "persistent://my-property/my-ns/pattern-topic-2-" + key;
         String topicName3 = "persistent://my-property/my-ns/pattern-topic-3-" + key;
+		final String topicName4 = "non-persistent://my-property/my-ns/pattern-topic-4-" + key;
```
also here.

gordeevbr commented:
Thanks for reviewing this everyone.
@merlimat we expected this feature to work through Pulsar Java client. I guess to use that API we will have to write our own client logic that queries for non-persistent topics first. I will take a look.
@jiazhai I must have forgotten my IDE is configured to use 2 whitespaces.

merlimat commented:
>  @merlimat we expected this feature to work through Pulsar Java client. I guess to use that API we will have to write our own client logic that queries for non-persistent topics first. I will take a look.

Yes, I was meaning that in broker side we should use the client API (there are already several cases that already do that). Since a single broker will not have all the topics, and that non-persistent topics don't have additional metadata, the client API call is the only option to get the correct list of non-persistent topics.

An additional improvement could be to only make this call if we know the client is interested in non-persistent topics and skip it otherwise.

gordeevbr commented:
@merlimat @jiazhai I have updated this feature as you suggested.
I haven't yet added tests for this or fixed any old ones, just made sure it compiles. I will test this ASAP, just wanted to make sure this functionality is OK before I go any further.

This is what was done in the latest commits:

- Additional property "Mode" was added to "CommandGetTopicsOfNamespace" entity of Pulsar binary API via Protobuf. This is a enum that has 3 values: "Persistent", "Non-Persistent", or "All". Value "Persistent" was set to be the default one.

- Broker binary protocol server code was updated to retrieve topics from lookup based on this enum. More importantly, broker will attempt to retrieve a list of non-persistent topics from another cluster if necessary. A pool of Pulsar Clients was made for that purpose inside NamespaceService.

- I have attempted to do the same for HTTP APIs, but I have only done so for API V1. I am not sure what to do about API V2, as it seems to use a somewhat more complicated functionality.

- Client code was updated to support this. When building a consumer it is now possible to specify to which kind of topics you want to subscribe when using pattern subscription.

Will be glad to receive your feedback.

sijie commented:
@merlimat @jiazhai can you review this PR and give @gordeevbr some feedback?

sijie commented:
ping @merlimat and @jiazhai . lets give some feedback to @gordeevbr , so that we can move forward with the change here.

jiazhai commented:
retest this please

jiazhai commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v1/Namespaces.java:
```diff
@@ -18,46 +18,31 @@
-import org.apache.pulsar.common.policies.data.AuthAction;
-import org.apache.pulsar.common.policies.data.BacklogQuota;
+import org.apache.pulsar.common.api.proto.PulsarApi.CommandGetTopicsOfNamespace.Mode;
+import org.apache.pulsar.common.policies.data.*;
```
we should avoid using `*` in import.

jiazhai commented on pulsar-broker/src/test/java/org/apache/pulsar/client/impl/PatternTopicsConsumerImplTest.java:
```diff
@@ -207,15 +213,106 @@ public void testBinaryProtoToGetTopicsOfNamespace() throws Exception {
+        List<String> topics = ((PatternMultiTopicsConsumerImpl<?>) consumer).getPartitionedTopics();
+        List<ConsumerImpl<byte[]>> consumers = ((PatternMultiTopicsConsumerImpl<byte[]>) consumer).getConsumers();
+
+        assertEquals(topics.size(), 1);
```
Test failed at here.  
Before patternconsumer created,  the producer had been created. By design, it should have 7 topics (1+2+3+1)?

gordeevbr commented:
@jiazhai @merlimat @sijie 

I have fixed old tests that were failing, added some new tests to test suites where it was possible, updated features to be consistent with HTTP Lookup, fixed all found issues, and updated with master branch.

There's not much new in these commits, mostly fixes and tests.

I think it should be ready now. I am, of course, ready to update this PR if deemed necessary.

sijie commented:
@gordeevbr cool, thanks. I removed "[wip]". so other people know this PR is ready to review.

gordeevbr commented:
Would you mind reruning these tests? I'm quite sure some of these were working just now.

sijie commented:
run java8 tests

sijie commented:
retest this please

sijie commented:
run java8 tests

gordeevbr commented:
Great. Thank you very much!

VijayRohra commented:
This is not working for me in Pulsar 2.11.1. I am able to receive messages from non-persistent topics belonging to single broker only.
